### PR TITLE
KAFKA-8803: [Streams fix] retry for initTxn

### PR DESCRIPTION
**Pull request by abbccdda**
Currently the initTransactions() API on StreamTask will throw uncaught timeout exception. It is not a desirable outcome as we normally would expect to retry upon timeout exception as this could indicate some temporary server side unavailability. Failing immediately is not ideal.

The fix is that if we hit non-fatal exception,  stream task keeps retrying `initTransactions` until succeed.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by vvcephei** [comment]
Hey @abbccdda , is this already covered by the Producer timeout configs (such as `max.block.ms`)?

There seems to be an ongoing debate in Streams about whether it's good to just set super-high timeouts on the Client configurations vs. retrying these operations in Streams. At the moment, though, I believe the accepted approach is just to increase the Client timeouts.

**Comment by abbccdda** [comment]
Another angle is whether we suppose a server's unavailability should penalize client: for txn initializing case specifically, we could have exception either due to network, broker failure, etc. Since we couldn't bring back a stream thread, hard shutdown is probably too much

**Comment by vvcephei** [comment]
Agreed, that's a good question, but if the timeout is high enough, then there would be no exception, and therefore no shutdown, right?

I believe the client already retries internally, so it should be able to recover from stuff like hung TCP connections, etc, albeit within the timeout you provide.

**Comment by abbccdda** [comment]
The trick is that time for server unavailability is unpredictable, so it's also hard for clients to pick an optimal timeout IMHO.

**Comment by vvcephei** [comment]
This is a good reason to change the way that Streams deals with these exceptions, but we should also make that design change holistically. Probably we should just make a new Jira, have a preliminary discussion there, and then proceed to a KIP discussion.

**Comment by abbccdda** [comment]
That sounds good to me.

**Comment by bbejeck** [comment]
I think both of you (@vvcephei and @abbccdda ) have good points.

While it's accurate users can increase the `max.block.ms` config, but max block configuration is used elsewhere.  So increasing max block to fix possible `initProducerId` issues, it could create other problems.

I'm not opposed to having some retries, **_but_** I don't think they should be unlimited.  
This goes for timeout exceptions in general.  I agree with @vvcephei. We should address this issue holistically and have a basic approach for all timeout exceptions.  Dying on the first one is not ideal, but letting streams hang around forever if there is a prolonged network partition or the broker is down and not coming back isn't good either.

**Comment by abbccdda** [comment]
@bbejeck Just got a ticket to track general improvements: https://issues.apache.org/jira/browse/KAFKA-9274

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 1]
```diff
@@ -906,21 +908,23 @@ long partitionTime(final TopicPartition partition) {
...
-                    "the broker was interrupted, or if similar circumstances arise. " +
-                    "You can increase producer parameter `max.block.ms` to increase this timeout.",
-                id,
-                retriable
-            );
-            throw new StreamsException(
-                format("%sFailed to initialize task %s due to timeout.", logPrefix, id),
-                retriable
-            );
+        while (true) {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:911]: Should we use `RETRIES_CONFIG` instead if trying forever?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java [hunk 2]
```diff
@@ -906,21 +908,23 @@ long partitionTime(final TopicPartition partition) {
...
+            try {
+                producer.initTransactions();
+                break;
+            } catch (final TimeoutException retriable) {
+                log.error(
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java:916]: Should we only log a WARN?

**Review by mjsax — COMMENTED** [review]
Maybe we should define a more general strategy/guidelines how we want to handle those cases? If a user never want to fail Streams if the brokers are not available, she should have corresponding configuration nobs -- @bbejeck brings up a very good point about the granularity of configs.

Maybe it would be a helpful exercise, to go through the whole code base, list all broker calls we do, and what exceptions they might throw, and was (client) configs are there a user could set. This would help us to understand the dependencies better.

Btw: we should add a unit test for this change.

**Comment by rocketraman** [comment]
I also posted in the parent ticket, but I don't believe this is the right solution to the problem in KAFKA-8803. My comments:

> I already effectively have retries because my app, upon encountering this error, crashes with a fatal error and is restarted automatically, at which point it retries. If you look at the original posts above, you will see that some of my streams retried every couple of minutes over the course of multiple weeks to restart with no success, so adding a retry in the client isn't going to solve this problem.
> 
> You have to figure out what the underlying problem in the broker is that is causing the problem in the first place.

**Pull request closed without merging**